Adding mutating webhook admission plugin

caesarxuchao opened the pull request:
Ref #https://github.com/kubernetes/features/issues/492

I made a change to the API to plumb the `Patch` into the response. I'll rebase onto the actual API once https://github.com/kubernetes/kubernetes/pull/55829 is merged.

We should update the release notes to point to the user docs when we have any.

```release-note
Added mutation supports to admission webhooks.
```

TODO: 
- [ ] update test image to v6 after #55829 is merged
- [ ] rename the GenericAdmissionWebhook to ValidatingAdmissionWebhook
- [ ] reduce json marshal/unmarshal roundtrip: https://github.com/kubernetes/kubernetes/pull/54892#discussion_r151336838
- [ ] move the matching function to a common package that validating and mutating webhooks can both import.
- [ ] handle namespace GET failure gracefully for fail open webhook?

caesarxuchao commented:
/sig api-machinery

deads2k commented:
Could you start with a pull that gets the serialization of the data to the webhooks correct?  Copying the code that is broken gives us two places to fix it.

If you're worried about defining types, I'd also be happy to see those in a separate pull too.  You could also split out the client-building portion of the code to some re-useable chunk to avoid potential drift.

caesarxuchao commented:
@deads2k that's exactly what I planned to do :) I started this PR as a hotchpotch to get a sense of required changes.

deads2k commented:
> @deads2k that's exactly what I planned to do :) I started this PR as a hotchpotch to get a sense of required changes.

Oh, great.  Makes sense then.

Commit:
Adding the mutating webhook

Commit:
generated bazel

caesarxuchao commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission_test.go:
```diff
@@ -0,0 +1,649 @@
+/*
```
The tests is exactly the same as those in the `validating/admission_test.go`. I'm not sure if we should keep this file.

caesarxuchao commented:
/retest

caesarxuchao commented:
/retest

caesarxuchao commented:
/retest

lavalamp commented on cluster/gce/config-test.sh:
```diff
@@ -316,7 +316,7 @@ if [[ -n "${GCE_GLBC_IMAGE:-}" ]]; then
 
 if [[ -z "${KUBE_ADMISSION_CONTROL:-}" ]]; then
-  ADMISSION_CONTROL="Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,PodPreset,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,Priority"
+  ADMISSION_CONTROL="MutatingAdmissionWebhook,Initializers,NamespaceLifecycle,LimitRanger,ServiceAccount,PersistentVolumeLabel,PodPreset,DefaultStorageClass,DefaultTolerationSeconds,NodeRestriction,Priority"
```
It's on my list to document the correct order for this. I don't know what else I will say, but I'm pretty sure initializers should go before this.

caesarxuchao replied:
Discussed offline, we perhaps want to disallow mutating webhooks to change initializer list to begin with, so I'll keep the current order in this PR.

deads2k replied:
Ah, I wondered where that question came from.  I think having initializers last is reasonable, but I think we need to consider a case where a deployer thinks they have a better way to wire initializers.  In such a case, they should be able to disable the compiled in initializer plugin and provide their own via webhook.  I'd probably say that our initializer plugin will always zero the list and stomp it to whatever we want, but a user *could* disable it and provide their own if they wished.

caesarxuchao replied:
Sent https://github.com/kubernetes/kubernetes/pull/55976.

lavalamp commented on cmd/kube-apiserver/app/options/options_test.go:
```diff
@@ -104,8 +104,8 @@ func TestAddFlags(t *testing.T) {
-			RecommendedPluginOrder: []string{"NamespaceLifecycle", "Initializers", "GenericAdmissionWebhook"},
-			DefaultOffPlugins:      []string{"Initializers", "GenericAdmissionWebhook"},
+			RecommendedPluginOrder: []string{"MutatingAdmissionWebhook", "NamespaceLifecycle", "Initializers", "GenericAdmissionWebhook"},
+			DefaultOffPlugins:      []string{"MutatingAdmissionWebhook", "Initializers", "GenericAdmissionWebhook"},
```
Generic needs to be renamed validating.

caesarxuchao replied:
Already listed in the todo :) Will do that in antoher PR.

lavalamp commented on cmd/kube-apiserver/app/options/options_test.go:
```diff
@@ -104,8 +104,8 @@ func TestAddFlags(t *testing.T) {
 		Admission: &apiserveroptions.AdmissionOptions{
-			RecommendedPluginOrder: []string{"NamespaceLifecycle", "Initializers", "GenericAdmissionWebhook"},
-			DefaultOffPlugins:      []string{"Initializers", "GenericAdmissionWebhook"},
+			RecommendedPluginOrder: []string{"MutatingAdmissionWebhook", "NamespaceLifecycle", "Initializers", "GenericAdmissionWebhook"},
```
Initializers should be first.

lavalamp replied:
Changed my mine, as discussed. :)

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/doc.go:
```diff
@@ -0,0 +1,18 @@
+limitations under the License.
+*/
+
+// Package mutating checks mutating webhooks for configured operation admission
```
maybe "makes calls to mutating webhooks during the admission process"?

caesarxuchao replied:
Done.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/validating/admission.go:
```diff
@@ -248,7 +248,6 @@ func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) error {
 				if ignoreClientCallFailures {
 					glog.Warningf("Failed calling webhook, failing open %v: %v", hook.Name, callErr)
 					utilruntime.HandleError(callErr)
-					// Since we are failing open to begin with, we do not send an error down the channel
```
Are we failing open? If so, leave a comment, if not, perhaps we should send an error down the channel?

caesarxuchao replied:
fail open v.s. fail closed is captured by `ignoreClientCallFailures`.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/server/options/admission.go:
```diff
@@ -56,8 +57,8 @@ func NewAdmissionOptions() *AdmissionOptions {
 		PluginNames:            []string{},
-		RecommendedPluginOrder: []string{lifecycle.PluginName, initialization.PluginName, validatingwebhook.PluginName},
-		DefaultOffPlugins:      []string{initialization.PluginName, validatingwebhook.PluginName},
+		RecommendedPluginOrder: []string{mutatingwebhook.PluginName, lifecycle.PluginName, initialization.PluginName, validatingwebhook.PluginName},
```
same comment about order

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission.go:
```diff
@@ -0,0 +1,316 @@
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+	objJS, err := runtime.Encode(a.jsonSerializer, attr.Object)
```
// TODO: if we have multiple mutating webhooks, we can remember the json instead of encoding and decoding for each one.

caesarxuchao replied:
I actually had it implemented in my prototype. I'll add the code back in another PR.

caesarxuchao replied:
Added.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission.go:
```diff
@@ -0,0 +1,316 @@
+}
+
+func (a *MutatingWebhook) shouldCallHook(h *v1alpha1.Webhook, attr admission.Attributes) (bool, *apierrors.StatusError) {
+	var matches bool
```
// TODO: factor into a common place along with the validating webhook version.

sttts replied:
+1

caesarxuchao replied:
Done.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/versioned/conversion.go:
```diff
@@ -28,6 +28,17 @@ type Convertor struct {
+func (c Convertor) Convert(in runtime.Object, out runtime.Object) error {
+	// for custom resources, because ConvertToGVK reuses the passed in object as
+	// the output, so they can skip the conversion here.
+	if in == out {
```
This is strictly an optimization and not a correctness thing, right? Can you add this detail to the comment?

caesarxuchao replied:
It's also for correctness. Currently the conversion stack will set the object to empty if we send the same runtime.Object as `in` and `out` to `Convert(in, out, context)`.

Will add that into the comment.

caesarxuchao replied:
Done.

lavalamp reviewed (commented):
I probably need to make another pass but these are my first thoughts.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission.go:
```diff
@@ -0,0 +1,316 @@
+		return fmt.Errorf("the MutatingWebhook admission plugin's jsonSerializer is not properly setup")
+	}
+	if err := a.namespaceMatcher.Validate(); err != nil {
+		return fmt.Errorf("the MutatingWebhook.namespaceMatcher is not properly setup: %v", err)
```
-the

caesarxuchao replied:
Done here and in the validating webhook.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission.go:
```diff
@@ -0,0 +1,316 @@
+	if err := a.convertor.Validate(); err != nil {
+		return fmt.Errorf("the MutatingWebhook.convertor is not properly setup: %v", err)
+	}
+	go a.hookSource.Run(wait.NeverStop)
```
this call hurts my eyes inside the validation. Looks like we need an `admission/Interface.Run`.

caesarxuchao replied:
> this call hurts my eyes inside the validation

we all agreed on this since 1.7 :)

sttts replied:
Which makes it even worse that we did not manage to fix it meanwhile ;-)

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission.go:
```diff
@@ -0,0 +1,316 @@
+	for _, r := range h.Rules {
+		m := rules.Matcher{Rule: r, Attr: attr}
+		if m.Matches() {
+			matches = true
```
commented in another PR before: return early here.

caesarxuchao replied:
It still has to pass the namespaceSelector's check. (maybe in the other PR the namesapceSelector's check is before the rule matching, the order was changed recently)

lavalamp replied:
at least break?

caesarxuchao replied:
There is a break after this line :)

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission.go:
```diff
@@ -0,0 +1,316 @@
+}
+
+// note that callHook updates attr
+func (a *MutatingWebhook) callHook(ctx context.Context, h *v1alpha1.Webhook, attr versioned.Attributes) error {
```
callAttrMutatingHook

caesarxuchao replied:
Done.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/certs_test.go:
```diff
@@ -0,0 +1,216 @@
+/*
```
please share those files.

caesarxuchao replied:
Already done in another commit :)

sttts commented:
Mostly as expected. Would like to see less code duplication.

caesarxuchao commented:
Got a strange verify-godep.sh failure,

```
I1116 05:28:40.450] Verifying hack/make-rules/../../hack/verify-godeps.sh
I1116 05:28:40.475] Checking for 'Godeps/' changes against 'bootstrap-upstream/master'
I1116 05:28:42.343] +++ [1116 05:28:42] Restoring kubernetes godeps
I1116 05:28:42.420] +++ [1116 05:28:42] Downloading dependencies - this might take a while
I1116 05:31:19.456] +++ [1116 05:31:19] Done
I1116 05:31:19.702] +++ [1116 05:31:19] Ensuring prereqs
I1116 05:31:19.777] +++ [1116 05:31:19] Running godep save - this might take a while
W1116 05:36:48.952] godep: no buildable Go source files in /tmp/gopath.TTnKPA/src/k8s.io/kubernetes
I1116 05:36:49.141] Removing /tmp/gopath.TTnKPA
I1116 05:36:51.780] FAILED   hack/make-rules/../../hack/verify-godeps.sh	491s
```

caesarxuchao commented:
/retest

caesarxuchao commented:
/retest

caesarxuchao commented on test/images/webhook/Makefile:
```diff
@@ -14,7 +14,7 @@
 build:
 	CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -o webhook .
-	docker build --no-cache -t gcr.io/kubernetes-e2e-test-images/k8s-sample-admission-webhook-amd64:1.8v3 .
+	docker build --no-cache -t gcr.io/kubernetes-e2e-test-images/k8s-sample-admission-webhook-amd64:1.8v4 .
```
TODO: update it to v6 after #55829 is merged.

caesarxuchao commented:
Still debugging the godep error.

lavalamp commented on garbage.go:
```diff
@@ -0,0 +1,3 @@
+package garbage
```
oh man, I laughed at this way harder than I should have

lavalamp commented:
The staging godep script is failing with a similar error. hm.

```
godep: no buildable Go source files in /tmp.k8s/tmp.Ue7oqEKRXx/go/src/k8s.io/apiserver
```

lavalamp commented:
Any chance it is a godep or go version change?

caesarxuchao commented:
After the `godep, i'm giving you what you want` commit, it passed.

I then removed the /garbage.go, and added `-v` to `godep save`, now the pre-submit is still running.

And strangely, it passed locally..... maybe the `-v` is the trick to make godep happy?

lavalamp commented:
I don't know why *this* PR triggers this. It's very strange. Other PRs are fine...

caesarxuchao commented:
SUCCESS  hack/make-rules/../../hack/verify-godeps.sh	694s

I'll remove the `-v` flag and try again.

Commit:
Rename the testdata package to testcerts.

`godep save` somehow fails if there is a testdata go package. See
https://github.com/kubernetes/kubernetes/pull/54892#issuecomment-345035489

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/testdata/gencerts.sh:
```diff
@@ -1,109 +0,0 @@
-#!/bin/bash
```
Did you delete this file on purpose?

caesarxuchao replied:
Oops. Added back.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission.go:
```diff
@@ -0,0 +1,318 @@
+	for i := range hooks {
+		call, err := a.shouldCallHook(&hooks[i], attr)
+		if err != nil {
+			return err
```
The only way we get here is if the namespace selector failed to parse, I think?

Should never happen. But maybe we shouldn't fail the whole request if the hook is also a fail-open hook?

Maybe something to think about in a follow-up.

lavalamp reviewed (commented):
/lgtm
/approve

Commit:
fix webhook e2e test cleanup

caesarxuchao commented:
/retest

cheftako commented on test/e2e/apimachinery/webhook.go:
```diff
@@ -682,9 +687,6 @@ func cleanWebhookTest(client clientset.Interface, namespaceName string) {
 	_ = client.ExtensionsV1beta1().Deployments(namespaceName).Delete(deploymentName, nil)
 	_ = client.CoreV1().Secrets(namespaceName).Delete(secretName, nil)
 	_ = client.RbacV1beta1().RoleBindings("kube-system").Delete(roleBindingName, nil)
-	_ = client.CoreV1().ConfigMaps(skippedNamespaceName).Delete(disallowedConfigMapName, nil)
```
You need to remove the delete webhook configurations from the cleanWebhookTest

caesarxuchao replied:
Done.

caesarxuchao commented:
I'll add the lgtm label after 55829 gets merged.

caesarxuchao commented:
/retest

caesarxuchao commented:
CI out of resource.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/admission.go:
```diff
@@ -0,0 +1,318 @@
+func (a *MutatingWebhook) SetScheme(scheme *runtime.Scheme) {
+	if scheme != nil {
+		a.clientManager.SetNegotiatedSerializer(serializer.NegotiatedSerializerWrapper(runtime.SerializerInfo{
+			Serializer: serializer.NewCodecFactory(scheme).LegacyCodec(admissionv1alpha1.SchemeGroupVersion),
```
Why is this here?  NegotiatedSerializerWrapper is not intended to be used.  Why aren't you passing the scheme's codec factory?

caesarxuchao replied:
Let me check that in a follow up.
